File: backend/src/services/auth.service.js
const bcrypt = require('bcryptjs');
const { AppError } = require('../middleware/error-handler');
const { findUserByEmail, createUser, updateLastLogin, findUserById } = require('../repositories/user.repository');
const { signToken } = require('../utils/jwt');

async function signup({ name, email, password }) {
  const normalizedEmail = email.trim().toLowerCase();
  const existingUser = await findUserByEmail(normalizedEmail);

  if (existingUser) {
    throw new AppError(409, 'EMAIL_ALREADY_EXISTS', 'An account with this email already exists.');
  }

  const passwordHash = await bcrypt.hash(password, 12);
  const user = await createUser({
    name: name.trim(),
    email: normalizedEmail,
    passwordHash,
  });

  return {
    user: {
      id: user.id,
      name: user.name,
      email: user.email,
      role: user.role,
    },
  };
}

async function login({ email, password }) {
  const normalizedEmail = email.trim().toLowerCase();
  const user = await findUserByEmail(normalizedEmail);

  if (!user || !(await bcrypt.compare(password, user.password_hash))) {
    throw new AppError(401, 'INVALID_CREDENTIALS', 'Invalid email or password.');
  }

  if (!user.is_active) {
    throw new AppError(403, 'ACCOUNT_INACTIVE', 'This account is inactive.');
  }

  await updateLastLogin(user.id);

  const token = signToken({ id: user.id, role: user.role });

  return {
    token,
    user: {
      id: user.id,
      name: user.name,
      email: user.email,
      role: user.role,
    },
  };
}

async function getCurrentUser(userId) {
  const user = await findUserById(userId);

  if (!user) {
    throw new AppError(401, 'AUTH_REQUIRED', 'Authentication required.');
  }

  if (!user.is_active) {
    throw new AppError(403, 'ACCOUNT_INACTIVE', 'This account is inactive.');
  }

  return {
    id: user.id,
    name: user.name,
    email: user.email,
    role: user.role,
  };
}

module.exports = {
  signup,
  login,
  getCurrentUser,
};

File: backend/services/auth_service.py
from sqlalchemy.orm import Session

from core.security import create_access_token, hash_password, verify_password
from exceptions.app_exceptions import (
    AccountInactiveError,
    EmailAlreadyExistsError,
    InvalidCredentialsError,
)
from models.user import User
from repositories.user_repository import UserRepository
from schemas.auth import LoginRequest, SignupRequest


class AuthService:
    def __init__(self, db: Session) -> None:
        self.repository = UserRepository(db)

    def signup(self, payload: SignupRequest) -> tuple[User, str]:
        normalized_email = payload.email.strip().lower()

        if self.repository.find_by_email(normalized_email):
            raise EmailAlreadyExistsError()

        password_hash = hash_password(payload.password)
        user = self.repository.create(
            name=payload.name.strip(),
            email=normalized_email,
            password_hash=password_hash,
            role="PICKER",
        )

        token = create_access_token(subject=str(user.id), role=user.role)
        return user, token

    def login(self, payload: LoginRequest) -> tuple[User, str]:
        normalized_email = payload.email.strip().lower()
        user = self.repository.find_by_email(normalized_email)

        if not user or not verify_password(payload.password, user.password_hash):
            raise InvalidCredentialsError()

        if not user.is_active:
            raise AccountInactiveError()

        self.repository.update_last_login(user)

        token = create_access_token(subject=str(user.id), role=user.role)
        return user, token
